Add synthetic example and documentation

hemmelig opened the pull request:
<!--
Thank your for contributing to QuakeMigrate!
Please fill out the sections below.
-->

## What is the purpose of this Pull Request? 
Add the synthetic example used in the QuakeMigrate paper to the examples directory and add online documentation describing each stage.

### Relevant Issues
There are no relevant Issues, this is a general addition that expands on the available documentation.

## Test cases
I have built the documentation locally using sphinx-build, which should hopefully replicate the building of the documentation over on readthedocs. The example itself runs (though the base branch from which I have branched off was not master but a separate branch in which some recently identified bugs had been fixed).

The example could be added as a further benchmark—thoughts? It doesn't really add too much different to the existing examples that are run to produce the benchmarks, but it equally couldn't hurt.

- [x] All examples run without any new warnings

### System details
This example was written and tested using Python 3.8 and 3.11, on macOS 12.5.1 running on a MacBook Pro with M1 chip.

### Final checklist
- [x] Correct base branch selected?
    - `master` (if a bugfix/patch update)
    - `version/vX.X.X` if a new significant feature (discuss with developers)
- [x] All tests still pass.
- [x] Any new or changed features are fully documented.
- [ ] Add proper links to other sections of the documentation where indicated (would require some of them to be written, is all).

TomWinder reviewed (commented):
This is sweet. Awesome.

I've speed-read through and noticed a couple of very minor typos. Only other comment is I think adding a subsection in detect labouring the point that the LUT determines the search grid (in space) - as well as the timespan etc that you already cover - would be worth doing.

The ability to use this framework to play around with other waveforms / network configs / etc etc etc is super powerful. Nice.

*EDIT: also yes - think it doesn't harm to add as a benchmark. Could be expanded in future to provide a simple test (or basis for unit tests) of different onset functions etc etc.

hemmelig commented:
Yeah, having this form a base dataset for testing different onset functions was one of the things I thought getting this integrated would be super useful for. The other one I realised out of the blue last week was—with this we might be able to sweep over STA and LTA values for a range of synthetics with different dominant frequencies, and get some concrete sense of sensible STA windows given a frequency range of interest, and how many times longer the LTA should be?

hemmelig commented:
I think we could add in the outputting of the marginalised coalescence function here, as it makes sense to tie in with the synthetic example. I think also needs to be rebased now that a number of these changes have been merged into master, particularly the conflicts that are being highlighted for the example scripts.

TomWinder commented:
Yes, re. marginal coalescence i/o it doesn't make a difference to me. I have it written and committed in the branch where I've added the Askja example, but can be either or.

Reading back through the conflicts here vs changes that are relevant to this PR, am I right that basically it is just the final two commits (adding the synthetic example, plus the associated docs) that we want to keep? And if so, can rebase magically deal with that? Would be amazing to be good enough at git to just cut and paste commits to where they should be...

Might also want to consider if using any of the new signal transformations (and other changes) is useful/relevant here.

TomWinder commented:
Just to underline - output of marginal coalescence now included in #184

hemmelig commented:
Redirected this PR to version/v1.2.0 base branch. I'll look at what conflicts are being raised. You mentioned above that you'd spotted a typo or two in my docs, but I think otherwise this is quite close to getting merged in?

hemmelig commented:
Thought maybe after handling the merge conflicts we might be able to get git to recognise that we also don't need the first commit listed in this PR, but alas.

TomWinder commented:
Yep I can't remember what I saw before to be honest. Not sure I will get to look at this in any detail today, but will otherwise do so asap next week.

I've merged #186 into `version/v1.2.0` - unsure if that will create further conflicts or if git is at least able to work that one out.

TomWinder commented on examples/synthetic/2_synthetic_detect.py:
```diff
@@ -0,0 +1,63 @@
+lut.decimate([4, 4, 1], inplace=True)
+
+# --- Create new Onset ---
+onset = STALTAOnset(position="centred", sampling_rate=50)
```
I still generally use "classic" by default for detect, due to the much higher sensitivity of "centred" to data that contains sharp steps etc etc. As with many things this is based on gut feel and anecdotal experience (followed by good experience using "classic" - could of course be way better using "centred" and I just don't know about it..)

TomWinder replied:
Ah the accompanying docs on synthetic-locate do emphasise using "centred" instead of "classic", so I guess this was maybe not deliberate..

hemmelig replied:
Yes, likely just a copy-paste hangover. Will adjust.

TomWinder commented on examples/synthetic/4_synthetic_locate.py:
```diff
@@ -0,0 +1,76 @@
+lut = read_lut(lut_file=lut_out)
+
+# --- Create new Onset ---
+onset = STALTAOnset(position="centred", sampling_rate=100)
```
Deliberate to use 100 Hz for this, and 50 Hz in detect?

hemmelig replied:
Will check.

hemmelig replied:
Yeah, I think my logic was 'decimate in time for detect', but if it throws extra confusion in, we can just put both to 100 Hz. Was saving a second or something per run.

hemmelig replied:
(not even—0.3 seconds!)

TomWinder reviewed (changes requested):
Fantastico. I think all these comments are pretty minor and easy to do - but some certainly need checking.

General comment: all the run scripts are currently missing the spiel at the top to prevent numpy using all available threads.

The docs look great. I think the only thing that needs doing there is updates to reflect changes made to the scripts from my specific suggestions there. See your point about the many opportunities for links to as-yet-not-there "main sections". But that'll come, so I think it's fine to leave ~ as-is for the purposes of this PR.

TomWinder commented on examples/synthetic/0_synthetic_lut.py:
```diff
@@ -0,0 +1,73 @@
+
+
+# Build synthetic lookup table
+station_file = "./inputs/synthetic_stations.sta"
```
Actually turns out many of these are .txt. Eh. I would go with that or .csv purely so file explorers etc make it more accessible for someone to open/preview.

hemmelig replied:
Great point, yes. I think `.sta` has just become my default shorthand for station files, but definitely better to just use a standard.

TomWinder commented on examples/synthetic/0_synthetic_lut.py:
```diff
@@ -0,0 +1,73 @@
+stations["Name"] = [f"STA{i}" for i in range(10)]
+stations["Longitude"] = rng.uniform(low=-0.15, high=0.15, size=10)
+stations["Latitude"] = rng.uniform(low=-0.15, high=0.15, size=10)
+stations["Elevation"] = [-0.0] * 10
```
Later saw in documentation the "this could be not zero". But why -0.0, rather than 0.0?

hemmelig replied:
Honestly, not sure. They could certainly be varied a little, I just went with 0 for simplicity. I'll try some variation, maybe a random sampling from 0–300 m or so.

hemmelig commented on examples/synthetic/0_synthetic_lut.py:
```diff
@@ -0,0 +1,73 @@
+    units="km",
+    lon_0=0.0,
+    lat_0=0.0,
+    lat_1=-0.149,
+    lat_2=0.1491,
```
That's a really good note, I wasn't aware of that. Yes, I'll adjust accordingly.

hemmelig commented on examples/synthetic/inputs/synthetic_stations.sta:
True.

hemmelig commented on examples/synthetic/1_simulate_waveforms.py:
```diff
@@ -0,0 +1,43 @@
+import numpy as np
+from quakemigrate.io import read_lut
+
+import simulate as simulate
```
Again, not sure why, but I've updated it.

hemmelig commented on examples/synthetic/simulate/core.py:
```diff
@@ -0,0 +1,315 @@
+    onset_params: tuple[float, float] = [0.1, 1.5],
+    fs: float = 50.0,
+) -> tuple[np.ndarray, np.ndarray]:
+    """
+    Calculate the *centred* short-term average / long-term average onset function
+    as implemented in QuakeMigrate v1.0.0.as_integer_ratio
+
+                    value assigned here (last sample of LTA window)
+                          v
+    |---------------------||-------|
+               ^               ^
+           LTA window      STA window
+
+    Parameters
+    ----------
+    x: Input signal.
+    filt: Bandpass filter parameters: [low-cut (Hz), high-cut (Hz), corners]
+    onset_params: [STA window (secs), LTA window (secs)].
+    fs: Sampling rate.
+
+    Returns
+    -------
+    onset: Onset function.
+    filt_x: Filtered signal.
+
+    """
+
+    # Detrend
+    detrend(x, overwrite_data=True)
+    # Filter
+    b, a = butter_bandpass(filt[0], filt[1], fs, order=filt[2])
+    filt_x = lfilter(b, a, x)
+
+    # Calculate onset function
+    stw = int(round(onset_params[0] * fs))
+    ltw = int(round(onset_params[1] * fs))
+
+    onset = sta_lta_centred(detrend(filt_x), stw, ltw)
+
+    return onset, filt_x
```
Yep, you're right—I think these are inherited directly from the notebooks I was using to develop the synthetic waveforms for the example.

hemmelig commented on examples/synthetic/simulate/core.py:
```diff
@@ -0,0 +1,315 @@
+        """Instantiate GaussianDerivativeWavelet object."""
+
+        delta_T = 1 / frequency
+        sigma = delta_T / 6
```
Likely just a convenient factor that gave ~appropriate looking waveforms.

hemmelig commented on examples/synthetic/simulate/core.py:
```diff
@@ -0,0 +1,315 @@
+        )
+
+        # Roll wavelet so first motion is at ~midpoint of array
+        self.data = np.roll(data, int(sps * 0.5 / frequency) + 3) / max(data)
```
I think when I wasn't using it, I wasn't getting the first motion right into the middle of the array, again seems to have just been an artefact of development in a notebook where I was plotting it all up and playing with the numbers.

hemmelig commented on examples/synthetic/4_synthetic_locate.py:
```diff
@@ -0,0 +1,76 @@
+scan.plot_event_summary = True
+
+# --- Toggle writing of waveforms ---
+scan.write_cut_waveforms = True
+
+scan.write_marginal_coalescence = True
```
No, these were here to be used later to build the figures for the paper. Can drop from the example.

hemmelig commented on examples/synthetic/2_synthetic_detect.py:
```diff
@@ -0,0 +1,63 @@
+# --- Create new Onset ---
+onset = STALTAOnset(position="centred", sampling_rate=50)
+onset.phases = ["P", "S"]
+onset.bandpass_filters = {"P": [1, 10, 2], "S": [1, 10, 2]}
```
Yeah, I think I just used something that would be somewhat centred around the targeted 4 Hz for the simulated waveforms, but running with high cut of 14 seems to not make a major difference.

hemmelig commented on examples/synthetic/4_synthetic_locate.py:
```diff
@@ -0,0 +1,76 @@
+)
+
+# --- Set locate parameters ---
+scan.marginal_window = 1.5
```
Yeah, changed to 0.2 s (~approximate FWHM of the 1-D coalescence function), and it gives us a little bit tighter constraint on the location.

hemmelig commented on examples/synthetic/3_synthetic_trigger.py:
```diff
@@ -0,0 +1,42 @@
+
+# --- Static threshold ---
+trig.threshold_method = "static"
+trig.static_threshold = 3.5
```
Yep, 10 looks good.

hemmelig commented on examples/synthetic/simulate/__init__.py:
```diff
@@ -0,0 +1,15 @@
+
+"""
+
+from .core import GaussianDerivativeWavelet, simulate_waveforms, sta_lta_onset
```
Yep, gone.

hemmelig commented:
Revisiting has set off a few thoughts, mainly regarding a lot of potential wasted compute at the locate stage. I started bumping up the LUT resolution to see if my hunch about what the oscillating coalescence as you move towards the true maximum was due to—i.e., insufficient spatial resolution causing the peaks to aligned when they shouldn't. I was able to smooth the peak out nicely (node spacing of 0.2, 0.2, 0.1 for LUT), and drop the uncertainties in x/y/z. And then you're left with a wash of dark blue that, in reality, didn't need to be computed. We're already some of the way there with not letting the event location move too much from the triggered location, but maybe we should go a step further and just not compute anything outside a certain radius. We could then encourage people to build quite chunky LUTs (memory space allowing) and just snip parts out for each event location.

hemmelig commented:
I think I've addressed all comments and committed these adjustments here with notes.

TomWinder reviewed (dismissed):
Looks good. I've updated the docs to reflect these adjustments. Please have a last check through this then good to go from my side.

Curiosity I noticed when looking at the locate summary plot again - is it expected that the origin time comes out at 12:05:00.530 (i.e. half a second after 5 mins)?

hemmelig commented:
Yeah, I had noticed that too and tried to dig into it a bit.

- Not velocity model errors (i.e., insufficient spatial sampling causing the origin to fall sufficiently far from a 'node' such that you get travel time inaccuracies)
- ran each step of the simulation script and plotted the wavelet. The first motion is at ~0.0 s as expected, when you apply a centred onset and plot, it's correctly peaking around the first motion
- we are only adding up to 0.02 s of noise to the travel times

hemmelig commented:
Ok—there's a line in the simulate module that adds 2 / frequency to the `half_timespan`, which for the specific frequency target (4 Hz) = 0.5 seconds. The start time of the traces are still set as 00:00:00.000, so the origin time is then out. I've removed this line (again, been too long to know if it was there for a particular reason or not). Also updated docs, where it appears. Still have an origin time reported as 0.16 s out, but this could be due to compounding errors on the travel time move outs (up to 0.02 s per trace, not centred, I could maybe also adjust that).

hemmelig commented:
And if you use a high resolution LUT, it drops to 12:05:00.02 for origin time (and really small spatial uncertainties!). I did have to use a 0.4 s marginal window to get it to locate, rather than 0.2 s.

I'll push the changes, then if you're happy to re-approve I will merge it.

TomWinder reviewed (approved):
Yeh it's just in the nook of the decimation-induced oscillation in the coalescence timeseries isn't it - makes sense. Looks like the synthetic hypocentre is pretty close to a grid edge in both latitude and longitude, and 0.03 s --> ~ 180 metres at 6km/s, so all well within the right ballpark, given everything is coupled. On reflection this was definitely worth addressing.

![image](https://github.com/user-attachments/assets/7d9f739a-4103-4d6f-b915-d313809a4f1f)

A great place to start with fiddling with boxcars and the like to remove this issue (at least in the origin time aspect - suspect the location will always be grid limited ---- BUT as with other stuff generally if you're grid limited you should be pretty happy). Satisfying to have all the info to chase it through 👍

TomWinder commented:
Note that eventually this should be added to `CHANGES.md`

hemmelig commented:
Added to CHANGES.md now, so we can deal with the merge clash properly when merging this PR (with other PRs that have also changed the CHANGES.md file). One final approve and I will merge.

hemmelig commented:
....another final approval (had to handle the merge conflict).

TomWinder commented:
🍾 🍾 🍾 🍾 🍾
